### PR TITLE
Mark options with performance impact in recommended options table

### PR DESCRIPTION
**Pull request by thomasnyman**
Based on the discussion during the C/C++ Compiler BP Guide meeting on [2023-04-12](https://docs.google.com/document/d/1x-fMIt6ZIV6SM-M29xEVKeB505GbDzSIVXg-8uF_FQU/edit#bookmark=id.q82jhtsux5i6) I've made an initial proposal for which options in the recommended options table should come with a note on performance impact in this PR. 

Please comment and discuss. Any additional information on when negative performance impact manifests would be good to also document in the more detailed description of the options.

### CODE REVIEW

**Comment by thomasnyman** [comment]
> Has there been any discussion on what features constitute being flagged about performance and which ones don't? In reality, almost all of the options listed have performance impacts. Is it 1 cycle? 10 cycles? 100 cycles? More?

Flagging options for potential performance impact was a todo item that was wished by several attendees at the [C/C++ Compiler BP Guide call on 2023-04-12](https://docs.google.com/document/d/1UClGUnOSkOH_wab6Lx43KUdkaK37L8sbWJ_GPZvc1YY/edit#bookmark=kix.k3onjgh7d07i). IIRC there wasn't really a clear criteria for which options should be flagged, but I took upon myself to take a initial attempt at this by flagging the options where the more detailed descriptions indicate conditions under which the options may incur an impact, even if that impact only manifests itself in certain contexts or edge cases.

Feedback on the specific options is of course welcome.

I don't entirely agree with the statement that "_almost all of the options_" have performance impact. The Table 1 in the guide is reserved for Warnings flags without an impact, and the linker flags that influence ELF attributes, not the generated code I think also fall under the category of options that never incur any impact on performance.

> I think it would make much more sense to simply state that many of the compiler options can have context specific performance impact that in general use is negligible.

I think this is a good suggestion and have added a note to that effect in https://github.com/ossf/wg-best-practices-os-developers/pull/144/commits/5538493e4f7fa38c47ec31f1815e72f0df4a79c6

File: docs/Compiler-Hardening-Guides/Compiler-Options-Hardening-Guide-for-C-and-C++.md [hunk 1]
```diff
@@ -89,7 +89,7 @@ The options in Table 2 can be categorized into two types:
 1) options that cause the compiler to augment the produced binary with run-time checks aimed to detect memory errors, and
 2) options that direct the compiler to adjust the properties of the generated binary or code to ensure the resulting binary is compatible with OS-enforced protection mechanisms.
 
-As these options have an impact on the binary produced by the compiler the effects of enabling any of the options in Table 2 must be validated through testing.
+As these options have an impact on the binary produced by the compiler the effects of enabling any of the options in Table 2 must be validated through testing. Many of these compiler options can have context-specific performance impact that in general use is negligible. The detailed descriptions of these options in this document elaborate on in which context non-neglible performance impact may manifest itself.
```
Comment (siddhesh) [inline docs/Compiler-Hardening-Guides/Compiler-Options-Hardening-Guide-for-C-and-C++.md:92]: Maybe something like:

"Security options can sometimes have a context-specific performance impact that in general use, is either negligible or the benefits far outweigh the overhead. The default descriptions of these options in this document elaborate on in which context non-neglible performance impact may manifest itself. If performance is a concern, then it is recommended to run benahcmarks for your application to understand the actual nature of the slowdown and weigh the benefits against the security protections."

**Comment by thomasnyman** [comment]
Resolved conflicts and incorporated feedback from the [C/C++ Compiler BP Guide call on 2023-06-21](https://docs.google.com/document/d/1UClGUnOSkOH_wab6Lx43KUdkaK37L8sbWJ_GPZvc1YY/edit#bookmark=kix.nsus8565fs6e).

**Comment by thomasnyman** [comment]
Merged as per discussion in [C/C++ Compiler Guide BP call 2023-08-02](https://docs.google.com/document/d/1UClGUnOSkOH_wab6Lx43KUdkaK37L8sbWJ_GPZvc1YY/edit#bookmark=id.82homkohrlln).

**Pull request merged**